tools/make_patch.c

```diff
@@ -0,0 +1,944 @@
+		if (!buffer) goto error;
+
+		switch (c) {
+		case '\n':
+			// A newline simply resets the line_pos
+			putc(c, output);
+			line_pos = 0;
+			break;
+
+		case '/':
+			// Check if we've encountered a comment
+			c = getc(file);
+			if (c != '*') {
+				putc('/', output);
+				line_pos++;
+				ungetc(c, file);
+				break;
+			}
+
+			// Simply gobble everything up until we hit the end
+			while ((c = getc(file)) != EOF) {
+				if (c == '*') {
+					if ((c = getc(file)) == EOF) break;
+					if (c == '/') break;
+				}
+			}
+
+			break;
+
+		case '{':
+			// Check if we've found two of them.
+			c = getc(file);
+			if (c != '{') {
+				putc('{', output);
+				line_pos++;
+				ungetc(c, file);
+				break;
+			}
+
+			// If we have, we store the contents before it ends into buffer
+			buffer_index = 0;
+			while ((c = getc(file)) != EOF) {
+				if (c == '}') {
+					if ((c = getc(file)) == EOF) break;
+					if (c == '}') break;
+				}
+
+				buffer[buffer_index++] = c;
+				buffer = expand_buffer(buffer, buffer_index);
+				if (!buffer) goto error;
+			}
+
+			buffer[buffer_index] = '\0';
+
+
+			patch->offset = 0;
+			interpret_command(buffer, currentpatch, symbols, patch, new_rom, orig_rom, output);
+			if (patch->offset) patch++;
+			patches = expand_buffer(patches, (uintptr_t)patch - (uintptr_t)patches);
+			if (!patches) goto error;
+
+			break;
+
+		case '[':
+			// End immediately if this is not the beginning of the line.
+			if (line_pos) {
+				putc(c, output);
+				line_pos++;
+				break;
+			}
+
+			// Try to read the label
+			putc(c, output);
+			buffer_index = 0;
+
+			while ((c = getc(file)) != EOF) {
+				putc(c, output);
+
+				if (c == ']') {
+					// If we're at the end, we can get the symbol for the label
+					buffer[buffer_index] = '\0';
+
+					// Translate spaces to underscores
+					for (char *p = buffer; *p != '\0'; p++) {
+						if (*p == ' ') *p = '_';
+					}
+
+					// Look for the symbol starting with prefix
+					char *searchlabel = malloc(strlen(prefix) + strlen(buffer) + 1);
+					strcpy(searchlabel, prefix);
+					strcat(searchlabel, buffer);
+					currentpatch = find_symbol(symbols, searchlabel);
+					if (!currentpatch) {
+						fprintf(stderr, "Error: Cannot find symbol: %s %s\n", prefix, buffer);
+					}
+
+					free(searchlabel);
+					// Skip until the next newline
+					while ((c = getc(file)) != EOF) {
+						putc(c, output);
+						if (c == '\n') break;
+					}
+					buffer_index = 0;
+					break;
+				}
+
+				buffer[buffer_index++] = c;
+			}
+
+			break;
+
+		default:
+			putc(c, output);
+			line_pos++;
+		}
+	}
+
+	patch->offset = 0;
+	free_buffer(buffer);
+	return patches;
+
+error:
+	if (patches) free_buffer(patches);
+	if (buffer) free_buffer(buffer);
+	return NULL;
+}
+
+int compare_patch(const void *_patch1, const void *_patch2)
+{
+	const struct patch *patch1 = _patch1;
+	const struct patch *patch2 = _patch2;
+	return patch1->offset - patch2->offset;
+}
```

Comment:
Unfortunately we have to use `void *` for the sake of `qsort`, but this needs fixing anyway to do signed subtraction):
```suggestion
int compare_patch(const void *patch1, const void *patch2)
{
	unsigned int patch1_offset = ((const struct patch *)patch1)->offset;
	unsigned int patch2_offset = ((const struct patch *)patch2)->offset;
	return patch1_offset > patch2_offset ? 1 : patch1_offset < patch2_offset ? -1 : 0;
}
```